File: Estruturas_de_Dados/abstratos/Grafo/Lista_Adjacente/grafo_lista_adjacente.py
# A única coisa que tenho a falar aqui é que, ao invés de self.graph
# ser uma matriz de arrays, aqui ele é um dicionário que guarda referencias
# a os nodes que tem conexões aos seus vizinhos
#
# Exemplo:
#
# {1: [2], 2: [1, 3], 3: [2]}
# - 1 ta conectado a 2
# - 2 ta conectado a 1 e 3
# - 3 ta conectado a 2

class Vertice:

    def __init__(self, data):
        
        self.data = data
    
    def __repr__(self):
        return f"{self.data}"

class Graph:

    def __init__(self):
        
        self.vertice_data = {}
        self.graph = {}
    
    def add_vertex(self, data_vertex):

        if data_vertex not in self.vertice_data.keys():
            vertice = Vertice(data_vertex)

            self.vertice_data[data_vertex] = vertice
            self.graph[vertice] = []

            return vertice
    
    def connect_vertices(self, vertice_1, vertice_2):

        if vertice_1 in self.vertice_data and vertice_2 in self.vertice_data:

            vertice_1 = self.vertice_data[vertice_1]
            vertice_2 = self.vertice_data[vertice_2]

            if vertice_1 == vertice_2:
                print("You cannot connect a vertice to itself!")
                return

            if vertice_1 not in self.graph[vertice_2]:
                self.graph[vertice_2].append(vertice_1)

            if vertice_2 not in self.graph[vertice_1]:
                self.graph[vertice_1].append(vertice_2)
        
        else:
            print(f"Cannot connect vertices, since either vertice {vertice_1} or vertice {vertice_2} do not exist")
            return

    def disconect_vertices(self, vertice_1, vertice_2):

        if vertice_1 in self.vertice_data and vertice_2 in self.vertice_data:

            vertice_1 = self.vertice_data[vertice_1]
            vertice_2 = self.vertice_data[vertice_2]

            if vertice_1 == vertice_2:
                print("You cannot disconect a vertice to itself!\n")
                return

            if vertice_1 in self.graph[vertice_2]:
                self.graph[vertice_2].remove(vertice_1)

            if vertice_2 in self.graph[vertice_1]:
                self.graph[vertice_1].remove(vertice_2)
        
        else:
            print(f"Cannot disconect vertices, since either vertice {vertice_1} or vertice {vertice_2} do not exist\n")
            return

    def print_depth_search(self, vertex = None, visited = {}):

        graph = self.graph

        if not vertex:
            
            vertex = list(self.vertice_data.values())[0]

        elif vertex in self.vertice_data:

            vertex = self.vertice_data[vertex]

        if not visited:
            visited[vertex] = True
            print(f"Performing Depth Search in node of data {vertex.data}:")

        print(vertex.data)

        for connected_vertice in self.graph[vertex]:
            
            if connected_vertice not in visited:    
                visited[connected_vertice] = True
                self.print_depth_search(connected_vertice, visited)
            elif visited[connected_vertice] == False:
                visited[connected_vertice] = True
                self.print_depth_search(connected_vertice, visited)
            
        #print(end="\n")
    
    
    def print_breadth_search(self, vertex = None):

        graph = self.graph
        queue = []
        visited = {}

        if not vertex:

            vertex = list(self.vertice_data.values())[0]

            queue.append(vertex)

            print(f"Performing Breadth Search in node of data {vertex.data}:")

        elif vertex in self.vertice_data:

            vertex = self.vertice_data[vertex]

            queue.append(vertex)            

            print(f"Performing Breadth Search in node of data {vertex.data}:")

        while queue:
            vertex = queue.pop(0)

            if vertex not in visited or visited[vertex] == False:

                print(vertex.data)
                visited[vertex] = True
                
                for connected_vert in self.graph[vertex]:
                    queue.append(connected_vert)
        
        print(end="\n")

    def print_graph(self):

        for vertice in self.graph.keys():
            print(f"Vertice {vertice}:", end = " ")
            print(", ".join(map(str, self.graph[vertice])))

        print(end="\n")
    

if __name__ == "__main__":

    grafo = Graph()

    for x in range(5):
        grafo.add_vertex(x)

    grafo.connect_vertices(1, 3)
    grafo.connect_vertices(2, 3)
    grafo.connect_vertices(3, 1)

    grafo.connect_vertices(4, 0)

    grafo.print_depth_search()
    grafo.print_depth_search(3)

    grafo.print_breadth_search()
    grafo.print_breadth_search(3)

    grafo.print_graph()
    print(grafo.graph)

File: Estruturas_de_Dados/abstratos/Grafo/Lista_Adjacente/GrafoListaAdjacente.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;

public class GrafoListaAdjacente {

    public static void main(String[] args) {

        Graph<Integer> grafo = new Graph<>();

        for (Integer x = 0; x <= 5; x++) {
            grafo.addVertex(x);};

        grafo.connectVertices(1, 3);
        grafo.connectVertices(2, 3);
        grafo.connectVertices(3, 1);
        grafo.connectVertices(4, 0);

        grafo.printGraph();

        System.out.println("\nDepth Search:");
        grafo.printDepthSearch();

        System.out.println("\nBreadth Search");
        grafo.printBreadthSearch();

    }
    
}

class Vertice<Thing> {

    Thing data;

    Vertice(Thing data) {
        
        this.data = data;

    }
}

class Graph<Thing> {

    HashMap<Vertice<Thing>, ArrayList<Vertice<Thing>>> graph = new HashMap<>();
    HashMap<Thing, Vertice<Thing>> verticeData = new HashMap<>();

    void addVertex(Thing dataVertex) {

        if (!this.verticeData.containsKey(dataVertex)) {
            Vertice<Thing> vertice = new Vertice<>(dataVertex);
            ArrayList<Vertice<Thing>> list = new ArrayList<>();

            this.verticeData.put(dataVertex, vertice);
            this.graph.put(vertice, list);
        }
    }

    void connectVertices(Thing verticeSourceData, Thing verticeDestinationData) {

        if (this.verticeData.containsKey(verticeSourceData) && this.verticeData.containsKey(verticeDestinationData)) {

            Vertice<Thing> verticeSource = this.verticeData.get(verticeSourceData);
            Vertice<Thing> verticeDestination = this.verticeData.get(verticeDestinationData);

            if (verticeSource.equals(verticeDestination)) {
                System.err.println("You cannot connect a vertice to itself");
                return;
            }

            if (this.graph.containsKey(verticeSource) && this.graph.containsKey(verticeDestination)) {

                ArrayList<Vertice<Thing>> tempList = this.graph.get(verticeDestination);
                
                if (!tempList.contains(verticeSource)) {
                tempList.add(verticeSource);}

                tempList = this.graph.get(verticeSource);
                
                if (!tempList.contains(verticeSource)) {
                tempList.add(verticeDestination);}
            }
            else {

                ArrayList<Vertice<Thing>> tempList = new ArrayList<>();

                this.graph.put(verticeSource, tempList);
                this.graph.put(verticeDestination, tempList);

                tempList = this.graph.get(verticeSource);
                tempList.add(verticeDestination);

                tempList = this.graph.get(verticeDestination);
                tempList.add(verticeSource);
            }

        }
    }

    void disconnectVertices(Thing verticeSourceData, Thing verticeDestinationData) {

        if (this.verticeData.containsKey(verticeSourceData) && this.verticeData.containsKey(verticeDestinationData)) {

            Vertice<Thing> verticeSource = this.verticeData.get(verticeSourceData);
            Vertice<Thing> verticeDestination = this.verticeData.get(verticeDestinationData);

            if (verticeSource.equals(verticeDestination)) {
                System.err.println("You cannot disconnect a vertice to itself");
                return;
            }

            ArrayList<Vertice<Thing>> sourceArray = this.graph.get(verticeSource);
            ArrayList<Vertice<Thing>> destinationArray = this.graph.get(verticeDestination);

            if (sourceArray.contains(verticeDestination) && destinationArray.contains(verticeSource)) {
            sourceArray.remove(verticeDestination);
            destinationArray.remove(verticeSource);
            } else {
                System.err.println("Can't disconnect nodes " + verticeSource.data + " and " + verticeDestination.data + " because they aren't connected");
            }

        } else {
            System.err.println("Cannot disconect vertices, since either vertice " + verticeSourceData + " or vertice " + verticeDestinationData + "do not exist");
        }
    }

    void printDepthSearch(Vertice<Thing> vertex, HashMap<Vertice<Thing>, Boolean> visited) {

        if (vertex == null) {

            LinkedList<Vertice<Thing>> tempArray = new LinkedList<>(this.verticeData.values());
            vertex = tempArray.getFirst();
            
        } else if (this.verticeData.containsValue(vertex)) {

            vertex = this.verticeData.get(vertex.data);
            
        }

        if (visited == null) {

            visited = new HashMap<>();
            visited.put(vertex, true);
            System.out.println("Performing Depth Search in node of data " + vertex.data);

        }

        System.out.println(vertex.data);

        for (Vertice<Thing> connectedVertice : this.graph.get(vertex)) {

            if (!visited.containsKey(connectedVertice)) {
                visited.put(connectedVertice, true);
                this.printDepthSearch(connectedVertice, visited);
            } else if (visited.get(connectedVertice) == false) {
                visited.put(connectedVertice, true);
                this.printDepthSearch(connectedVertice, visited);
            }
        }
    }

    void printBreadthSearch(Vertice<Thing> vertice) {

        LinkedList<Vertice<Thing>> queue = new LinkedList<>();
        HashMap<Vertice<Thing>, Boolean> visited = new HashMap<>();

        if (vertice == null) {

            LinkedList<Vertice<Thing>> tempList = new LinkedList<>(this.verticeData.values());
            vertice = tempList.getFirst();
            queue.addLast(vertice);
            System.out.println("Performing Breadth Search in node of data " + vertice.data + ":");

        } else if (this.verticeData.containsValue(vertice)) {

            vertice = this.verticeData.get(vertice.data);
            queue.addLast(vertice);
            System.out.println("Performing Breadth Search in node of data " + vertice.data + ":");

        }

        while (!queue.isEmpty()) {

            vertice = queue.getFirst();
            queue.removeFirst();

            System.out.println(vertice.data);
            visited.put(vertice, true);

            for (Vertice<Thing> connectedVert : this.graph.get(vertice)) {
                if (!visited.containsKey(connectedVert) || visited.get(connectedVert) == false) {
                queue.addLast(connectedVert);
            }
            }
        }

    }

    void printDepthSearch() {
        this.printDepthSearch(null, null);
    }

    void printBreadthSearch() {
        this.printBreadthSearch(null);
    }

    void printGraph() {

        boolean firstPrinted = false;
        System.out.println("\nPrinting Vertices Conections in Graph:");

        for (Vertice<Thing> vertice : this.graph.keySet()) {
            System.out.print("Vertice " + vertice.data + ": ");

            for (Vertice<Thing> connectedVertice : this.graph.get(vertice)) {

                if (firstPrinted) {
                System.out.print(", " + connectedVertice.data);
                } else {
                    System.out.print(connectedVertice.data);
                    firstPrinted = !firstPrinted;
                }
            }
            firstPrinted = !firstPrinted;
            System.out.print("\n");
        }
        System.out.print("\n");
    }

}